### PR TITLE
fix deadlock and use after close of tts_forwarder

### PR DESCRIPTION
**Pull request by theomonnom**
fix https://github.com/livekit/agents/issues/351 & function callings not working

### CODE REVIEW

**Comment by theomonnom** [comment]
Only the last commit is relevant

File: livekit-agents/livekit/agents/transcription/tts_forwarder.py [hunk 1]
```diff
@@ -190,38 +198,35 @@ def mark_text_segment_end(self) -> None:
         if self._closed:
             raise RuntimeError("mark_text_segment_end called after close")
 
-        # create new segment on "mark_text_segment_end"
+        # create new segment on "mark_text_segment_end" -> None:
```
Comment (davidzhao) [inline livekit-agents/livekit/agents/transcription/tts_forwarder.py:201]: ` -> None:` ?

File: livekit-agents/livekit/agents/transcription/tts_forwarder.py [hunk 2]
```diff
@@ -230,8 +235,10 @@ def _create_segment(self) -> _SegmentData:
             processed_hyphenes=0,
             ready_future=asyncio.Future(),
         )
+        self._next_segment_index += 1
+        return data
 
-    async def _forward(self, q: asyncio.Queue[rtc.TranscriptionSegment | None]):
+    async def _forward_co(self, q: asyncio.Queue[rtc.TranscriptionSegment | None]):
```
Comment (davidzhao) [inline livekit-agents/livekit/agents/transcription/tts_forwarder.py:241]: `_co` meaning coroutine? where they would be running in parallel?

File: livekit-agents/livekit/agents/voice_assistant/assistant.py [hunk 3]
```diff
@@ -327,7 +332,52 @@ async def _main_task(self) -> None:
...
+        #  - Decide when to validate the user speech (starting the agent answer)
+        speech_prob_avg = utils.MovingAverage(100)
+        speaking_avg_validation = utils.MovingAverage(150)
+        interruption_speaking_avg = utils.MovingAverage(
+            int(self._opts.int_speech_duration * 100)
+        )
+
+        interval_10ms = aio.interval(0.01)
+
+        vad_pw = 2.4  # TODO(theomonnom): should this be exposed?
```
Comment (davidzhao) [inline livekit-agents/livekit/agents/voice_assistant/assistant.py:348]: what does this value indicate?

**Review by davidzhao — APPROVED** [review]

File: livekit-agents/livekit/agents/voice_assistant/assistant.py (hunk 3 above)
Comment (theomonnom) [reply livekit-agents/livekit/agents/voice_assistant/assistant.py:348]: the amount of volume you decrease/increase based on the speech probability

Comment (theomonnom) [reply livekit-agents/livekit/agents/voice_assistant/assistant.py:348]: it's more like "how fast" you increase/decrease

File: livekit-agents/livekit/agents/transcription/tts_forwarder.py (hunk 2 above)
Comment (theomonnom) [reply livekit-agents/livekit/agents/transcription/tts_forwarder.py:241]: it means coroutine, it is running sequentially. Compared to a task that is spawned using asyncio.create_task.
The synthax I'm doing it here is just to separate both concept

**Pull request merged**